cpu/cc2538/periph/rtt.c

```diff
@@ -60,34 +71,70 @@ void rtt_init(void)
     rtt_poweron();
 }
 
+static inline uint32_t _rtt_get_counter(void)
+{
+    return ((SMWDTHROSC_ST0 & 0xFF)
+         | ((SMWDTHROSC_ST1 & 0xFF) << 8)
+         | ((SMWDTHROSC_ST2 & 0xFF) << 16)
+         | ((SMWDTHROSC_ST3 & 0xFF) << 24));
+}
+
 uint32_t rtt_get_counter(void)
 {
-    return ((SMWDTHROSC_ST0 & 0xFF) |
-            ((SMWDTHROSC_ST1 & 0xFF) << 8) |
-            ((SMWDTHROSC_ST2 & 0xFF) << 16) |
-            ((SMWDTHROSC_ST3 & 0xFF) << 24));
+    return _rtt_get_counter() - rtt_offset;
 }
 
-void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+void rtt_set_counter(uint32_t counter)
 {
-    assert(cb && !(alarm & ~RTT_MAX_VALUE));
+    rtt_alarm -= rtt_offset;
+    rtt_offset = _rtt_get_counter() + counter;
+    rtt_alarm += rtt_offset;
 
-    unsigned irq = irq_disable();
+    /* re-set the overflow callback */
+    if (overflow_cb) {
+        rtt_set_overflow_cb(overflow_cb, overflow_arg);
+    }
+}
 
-    /* set alarm value */
+static void _set_alarm(uint32_t alarm)
+{
     while (!(SMWDTHROSC_STLOAD & SMWDTHROSC_STLOAD_STLOAD_MASK)) {}
     SMWDTHROSC_ST3 = (alarm >> 24) & 0xFF;
     SMWDTHROSC_ST2 = (alarm >> 16) & 0xFF;
     SMWDTHROSC_ST1 = (alarm >>  8) & 0xFF;
     SMWDTHROSC_ST0 = alarm & 0xFF;
+}
+
+void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
+{
+    assert(cb && !(alarm & ~RTT_MAX_VALUE));
+
+    unsigned irq = irq_disable();
+    uint32_t now = _rtt_get_counter();
+
+    rtt_alarm = alarm + rtt_offset;
 
     /* set callback*/
     alarm_cb = cb;
     alarm_arg = arg;
 
+    DEBUG("rtt_set_alarm(%lu), alarm in %lu ticks, overflow in %lu ticks\n",
+          alarm, rtt_alarm - now, rtt_offset - now);
+
+    /* only set overflow alarm if it happens before the scheduled alarm */
+    if (overflow_cb == NULL || (rtt_offset - now >= rtt_alarm - now)) {
```

Comment:
```suggestion
    if (overflow_cb == NULL || (rtt_offset >= rtt_alarm )) {
```

Reply:
I wonder if this was to handle overflows, but can't remember.